KLayout DRC integration

bzhangflex opened the pull request:
Adds a new plugin that integrates KLayout's DRC.

### Features

- Run DRC checks on Tidy3D objects that support exporting to GDS (`Geometry`, `Structure`, `Simulation`)
- Basic DRC result parsing (check if DRC clean, gather violation counts)

### Main functions

1. `run_drc()`: Main function to run DRC on Tidy3D objects
2. `is_drc_clean()`: Check if a DRC result file has no violations
3. `count_drc_violations()`: Get a count of DRC violations by category

### Usage Example

```python
import tidy3d as td
from tidy3d.plugins.klayout.drc import run_drc

# Create a Tidy3D geometry
vertices = [(-2, 0), (-1, 1), (0, 0.5), (1, 1), (2, 0), (0, -1)]
geom = td.PolySlab(vertices=vertices, slab_bounds=(0, 0.22), axis=2)

# Run DRC
run_drc(
    geom,                # Tidy3D object
    "output.gds",       # Output GDS file
    "rules.drc",        # DRC rules file
    "results.lyrdb",    # Results file
    z=0.1,             # Additional GDS export parameters
    gds_layer=0,
    gds_dtype=0
)

# Check results
if is_drc_clean("results.lyrdb"):
    print("DRC passed!")
else:
    violations = count_drc_violations("results.lyrdb")
    print("DRC violations found:", violations)
```

### Documentation

A README is included with detailed instructions and usage examples. More documentation can be added elsewhere as needed (perhaps we add an example notebook as well)

<!-- greptile_comment -->

## Greptile Summary

This PR introduces a new plugin that integrates KLayout's Design Rule Check (DRC) functionality into Tidy3D, enabling geometric validation of designs before simulation. The implementation provides a clean, well-structured API for running DRC checks on both GDS files and Tidy3D objects that can be exported to GDS format.

The changes include a complete plugin architecture with configuration handling, result parsing, and comprehensive test coverage. Key additions include support for multi-polygon DRC markers, improved error handling, and clear documentation with usage examples.

## Confidence score: 4/5

1. This PR is quite safe to merge as it adds new functionality without modifying existing core behavior.
2. High confidence due to comprehensive test coverage and clean implementation, though minor API documentation improvements could be beneficial.
3. Files needing attention:
   - `tidy3d/plugins/klayout/drc/drc.py`: Core implementation file that wasn't fully visible in the context
   - `tidy3d/plugins/klayout/drc/README.md`: Documentation might need expansion for advanced use cases

<sub>11 files reviewed, no comments</sub>
<sub>[Edit PR Review Bot Settings](https://app.greptile.com/review/github) | [Greptile](https://greptile.com?utm_source=greptile_expert&utm_medium=github&utm_campaign=code_reviews&utm_content=tidy3d_2586)</sub>

<!-- /greptile_comment -->

bzhangflex committed:
feat(plugins): klayout drc plugin

yaugenst-flex commented on CHANGELOG.md:
```diff
@@ -25,6 +25,7 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 - Added heat sources with custom spatial dependence. It is now possible to add a `SpatialDataArray` as the `rate` in a `HeatSource`.
 - Added Transient Heat simulations. It is now possible to run transient Heat simulations. This can be done by specifying `analysis_spec` of `HeatChargeSimulation` object as `UnsteadyHeatAnalysis`. 
 - A `num_grid_cells` field to `WavePort`, which ensures that there are 5 grid cells across the port. Grid refinement can be disabled by passing `None` to `num_grid_cells`.
+- Added KLayout DRC plugin for running design rule checks in `plugins.klayout.drc`. Supports running DRC on `Geometry`, `Structure`, and `Simulation` objects. Added helper functions `is_drc_clean` and `count_drc_violations` for reading the results.
```
I'd probably reword this to just "Added KLayout plugin" and say that it has DRC features, since we might get around to adding some other features too eventually.

yaugenst-flex commented on tidy3d/plugins/klayout/__init__.py:
I'd suggest we pull in the most relevant user-facing functions from the plugin into the namespace here so a user can just `from tidy3d.plugins.klayout import run_drc`.

Or do you think it's better to keep this strictly separated?

bzhangflex replied:
Thanks, I added the user facing functions to the klayout namespace

yaugenst-flex commented on tidy3d/plugins/klayout/drc/README.md:
```diff
@@ -0,0 +1,93 @@
+# KLayout DRC Integration for Tidy3D
+
+This plugin provides integration between Tidy3D and KLayout's Design Rule Check (DRC) engine, allowing you to perform design rule checks on Tidy3D objects.
```
"module" instead of "plugin" here since this is technically a sub-module of the KLayout plugin?

yaugenst-flex commented on tidy3d/plugins/klayout/drc/README.md:
```diff
@@ -0,0 +1,93 @@
+
+2. The user must provide a KLayout DRC script that defines the source (input gds) using `source($gdsfile)` and the report (output result file) using `report("DRC results", $resultsfile)`. More below.
+
+## Features
```
I'd swap the "Prerequisites" and the "Features" section

yaugenst-flex commented on tidy3d/plugins/klayout/drc/README.md:
```diff
@@ -0,0 +1,93 @@
+
+The full path to the application should be displayed if KLayout has been added to the system PATH.
+
+2. The user must provide a KLayout DRC script that defines the source (input gds) using `source($gdsfile)` and the report (output result file) using `report("DRC results", $resultsfile)`. More below.
```
"More below" is fine albeit a bit redundant I guess. Might be better to link a specific section instead?

yaugenst-flex commented on tidy3d/plugins/klayout/drc/README.md:
This readme is awesome!

yaugenst-flex commented on tidy3d/plugins/klayout/drc/drc.py:
```diff
@@ -0,0 +1,291 @@
+from tidy3d.components.simulation import Simulation
+from tidy3d.components.structure import Structure
+
+## KLayout DRC running
```
Personally I don't like these commented "headers" in source files. If they seem necessary, then I'd say it's more of an indication that the module should be split up.

yaugenst-flex commented on tidy3d/plugins/klayout/drc/drc.py:
```diff
@@ -0,0 +1,291 @@
+        If the design rule file does not exist.
+    """
+    if not Path(designrulefile).is_file():
+        raise FileNotFoundError(f"Design rule file not found: '{designrulefile}'")
```
These two functions are nearly identical. I'd suggest refactoring into a single `_check_file_exists()` and use that within `run_drc_on_gds()` instead.

yaugenst-flex commented on tidy3d/plugins/klayout/drc/drc.py:
```diff
@@ -0,0 +1,291 @@
+        content = f.read()
+        if not re.search(r"source\(\s*\$gdsfile\s*\)", content):
+            raise SyntaxError(
+                "Design rule file is not formatted correctly. The GDS source must be loaded with 'source($gdsfile)'. Please refer to the README for more details."
```
Might be good to add a link to the README here, since it's not obvious where that file is if you've only pip-installed tidy3d.

yaugenst-flex commented on tidy3d/plugins/klayout/drc/drc.py:
```diff
@@ -0,0 +1,291 @@
+    _check_drcfile_format(designrulefile)
+
+    # check if klayout exists in the system path
+    _check_klayout_exists()
```
All these comments are redundant.

yaugenst-flex commented on tidy3d/plugins/klayout/drc/drc.py:
```diff
@@ -0,0 +1,291 @@
+    # check if klayout exists in the system path
+    _check_klayout_exists()
+
+    # run drc
```
ditto

yaugenst-flex commented on tidy3d/plugins/klayout/drc/drc.py:
```diff
@@ -0,0 +1,291 @@
+    td_object.to_gds_file(gdsfile, **to_gds_file_kwargs)
+
+    # run the DRC
+    run_drc_on_gds(gdsfile, designrulefile, resultsfile)
```
remove these comments

yaugenst-flex commented on tidy3d/plugins/klayout/drc/drc.py:
```diff
@@ -0,0 +1,291 @@
+        )
+
+
+def _check_drcfile_format(designrulefile: str | Path) -> None:
```
Note (here and thereafter): Pipe-style unions (`x | y`) were introduced in Python 3.10, but we currently still support 3.9, so all of these should be changed to `Union[x, y]` instead. The reason this doesn't break here is because all type hints are converted to forward references (strings) via `from __future__ import annotations` and there is no code in this PR that needs to actually _evaluate_ these types. This does happen in other parts of the codebase though, so to keep things consistent, these should be rewritten as `Union` until we drop support for 3.9 (which will be some time later this year when 3.9 is EOL).

yaugenst-flex reviewed (changes requested):
Thanks @bzhangflex this is great! Note the plugin should also be added to the docs here: https://github.com/flexcompute/tidy3d/tree/develop/docs/api/plugins

A question on the interface: Would it make sense to provide default paths for the `resultsfile` in `is_drc_clean()` and the `gdsfile` in `run_drc()`?

Also, currently there seems to be no direct way of running DRC and getting the results in a single call. The workflow is distinctly separated into a "run DRC" step, and then you explicitly load the results of that to do further analysis. I think this is a well-structured workflow, but as a user I'd probably prefer getting the results of the DRC right away (in some well-defined data structure), so essentially have `run_drc()` return something like a `DRCResults` data structure. To spin that thought a bit further, I think there should be a higher-level object-oriented interface (something that derives from `Tidy3dBaseModel`). Generally this would put things more in line with established conventions in the codebase but also allow us to serialize these things and e.g. run them in GUI (eventually, maybe). Note that I think this can all be implemented without changing the current functions, we'd just be wrapping them. Keeping the functional interface around is a good idea.

yaugenst-flex commented on tests/test_plugins/klayout/drc/test_drc.py:
```diff
@@ -0,0 +1,114 @@
+    assert is_drc_clean(filepath / "drc_results_widthonly_clean.lyrdb")
+
+
+def test_count_drc_violations():
+    """Test that count_drc_violations works"""
+    assert count_drc_violations(filepath / "drc_results_widthonly.lyrdb")["min_width"] == 2
```
@bzhangflex do you think this would be feasible to do without introducing new files? Otherwise we can ignore this.

bzhangflex replied:
added a few more tests

bzhangflex commented:
Thank you @yaugenst-flex for the detailed review!

> A question on the interface: Would it make sense to provide default paths for the `resultsfile` in `is_drc_clean()` and the `gdsfile` in `run_drc()`?

Hmm I'm not sure, the user might end up loading the wrong results unknowingly or overwrite an important gds. What do you think?

> Also, currently there seems to be no direct way of running DRC and getting the results in a single call. The workflow is distinctly separated into a "run DRC" step, and then you explicitly load the results of that to do further analysis. I think this is a well-structured workflow, but as a user I'd probably prefer getting the results of the DRC right away (in some well-defined data structure), so essentially have `run_drc()` return something like a `DRCResults` data structure. 

Very good point, will work towards that.

> To spin that thought a bit further, I think there should be a higher-level object-oriented interface (something that derives from `Tidy3dBaseModel`). Generally this would put things more in line with established conventions in the codebase but also allow us to serialize these things and e.g. run them in GUI (eventually, maybe). Note that I think this can all be implemented without changing the current functions, we'd just be wrapping them. Keeping the functional interface around is a good idea.

Would this be something like all of the sub methods get moved into a DRC object, and run_drc() remains a standalone function but creates and calls a DRC object within it?

yaugenst-flex commented:
> > A question on the interface: Would it make sense to provide default paths for the `resultsfile` in `is_drc_clean()` and the `gdsfile` in `run_drc()`?
> 
> Hmm I'm not sure, the user might end up loading the wrong results unknowingly or overwrite an important gds. What do you think?

Yes, but: We have a default file for simulation results too, i.e., if you don't specify a `path` in `web.run()`, you'll get a `simulation_data.hdf5`. If you think it's a problem, we can also skip adding a default for now. Generally I think it's better to reduce the number of required arguments, but we can leave it required for now and just see how it goes.

> > To spin that thought a bit further, I think there should be a higher-level object-oriented interface (something that derives from `Tidy3dBaseModel`). Generally this would put things more in line with established conventions in the codebase but also allow us to serialize these things and e.g. run them in GUI (eventually, maybe). Note that I think this can all be implemented without changing the current functions, we'd just be wrapping them. Keeping the functional interface around is a good idea.
> 
> Would this be something like all of the sub methods get moved into a DRC object, and run_drc() remains a standalone function but creates and calls a DRC object within it?

The other way round actually: Create a class that exposes these DRC methods, but just wraps the current functions. Somewhat similar to how it's done in the autograd plugin: https://github.com/flexcompute/tidy3d/blob/develop/tidy3d/plugins/autograd/invdes/filters.py

bzhangflex commented:
Thanks @yaugenst-flex, I've made quite a few changes. Now `run_drc()` returns a `DRCResults` data structure that includes all of the DRC violation markers.

Example usage:

```python
from tidy3d.plugins.klayout.drc import run_drc

# Create a simple polygon geometry
vertices = [(-2, 0), (-1, 1), (0, 0.5), (1, 1), (2, 0), (0, -1)]
geom = td.PolySlab(vertices=vertices, slab_bounds=(0, 0.22), axis=2)

# Run DRC and get results
results = run_drc(geom, "geom.gds", "test.drc", "drc_results.lyrdb", 
                 z=0.1, gds_layer=0, gds_dtype=0)

# Check if DRC passed
if results.is_drc_clean():
    print("DRC passed!")
else:
    print(results)  # Prints violation summary

# Violations are indexed by category
print(results['min_width'].violations)
```

yaugenst-flex commented on tidy3d/plugins/klayout/drc/drc.py:
```diff
@@ -0,0 +1,555 @@
+class EdgeMarker(Tidy3dBaseModel):
+    """A class for storing KLayout DRC edge marker results."""
+
+    edge: tuple[tuple[float, float], tuple[float, float]] = pd.Field(
```
Consider creating a type for `tuple[tuple[float, float], tuple[float, float]]`. this could also be a case for [`Coordinate2D`](https://github.com/flexcompute/tidy3d/blob/00b0330a49d122a2b484db226147ef7b8913d145/tidy3d/components/types.py#L193) in the base types.

yaugenst-flex commented on tidy3d/plugins/klayout/drc/drc.py:
```diff
@@ -0,0 +1,555 @@
+class PolygonMarker(Tidy3dBaseModel):
+    """A class for storing KLayout DRC polygon marker results."""
+
+    polygon: tuple[tuple[float, float], ...] = pd.Field(
```
Same here. Could be `tuple[Coordinate2D, ...]`

yaugenst-flex commented on tidy3d/plugins/klayout/drc/drc.py:
```diff
@@ -0,0 +1,555 @@
+        description="The number of DRC violations in the category.",
+    )
+    violations: list[Union[EdgeMarker, EdgePairMarker, PolygonMarker]] = pd.Field(
+        default=[], title="DRC Violations", description="List of DRC violations in the category."
```
No mutable defaults. We can either make the violations a `tuple` (any reason not to?) or use `default_factory=list`.

yaugenst-flex commented on tidy3d/plugins/klayout/drc/drc.py:
```diff
@@ -0,0 +1,555 @@
+class DRCResults(Tidy3dBaseModel):
+    """A class for loading and storing KLayout DRC results."""
+
+    resultsfile: Union[str, Path] = pd.Field(
```
You could leave this as `Path`, I _think_ that would still work (and automatically coerce strings into `Path`s).

yaugenst-flex commented on tidy3d/plugins/klayout/drc/drc.py:
```diff
@@ -0,0 +1,555 @@
+    )
+    clean: bool = pd.Field(
+        title="DRC Clean", description="Whether the DRC is clean (no violations)."
+    )
```
It seems to me like this should be a property that gets calculated based on `violations` instead of a static field?

yaugenst-flex commented on tidy3d/plugins/klayout/drc/drc.py:
```diff
@@ -0,0 +1,555 @@
+        for category in xmltree.getroot().findall(".//categories/category/name"):
+            violations[category.text] = DRCViolation(category=category.text)
+
+        # Prase violations
```
typo: `Prase` -> `Parse`

yaugenst-flex commented on tidy3d/plugins/klayout/drc/drc.py:
```diff
@@ -0,0 +1,555 @@
+        dict[str, int]
+            A dictionary of violation counts for each category.
+        """
+        return {category: violation.count for category, violation in self.violations.items()}
```
Since this doesn't take arguments consider turning into a (cached) `violation_counts` property?

yaugenst-flex commented on tidy3d/plugins/klayout/drc/drc.py:
```diff
@@ -0,0 +1,555 @@
+        description="The path to the GDS file to write the Tidy3D object to.",
+    )
+    designrulefile: Union[str, Path] = pd.Field(
+        default=None,
```
Mark `Optional`.

bzhangflex replied:
Actually it's not optional, I should probably remove the default

yaugenst-flex commented on tidy3d/plugins/klayout/drc/drc.py:
```diff
@@ -0,0 +1,555 @@
+
+        Notes
+        -----
+        The user is encouraged to use the function ``run_drc`` instead of this class.
```
Why?

yaugenst-flex commented on tidy3d/plugins/klayout/drc/drc.py:
Nice, I think this is going in the right direction. However, as discussed, I think the abstraction hierarchy is inverted. The pure functions (i.e., the functional API) is more fundamental than the object-oriented one. Hence, I think the class should wrap the functions, and not the other way round. Basically `DRCRunner` should be a higher-level of abstraction then the `run_drc()` functions that adds additional convenience methods.

bzhangflex replied:
The code has been refactored, let me know how it looks.

yaugenst-flex commented on tidy3d/plugins/klayout/drc/drc.py:
```diff
@@ -0,0 +1,555 @@
+
+    def __str__(self) -> str:
+        """Get a nice string representation of the DRC results."""
+        str = "DRC results summary\n"
```
Don't shadow built-in `str`

yaugenst-flex commented on tidy3d/plugins/klayout/drc/drc.py:
```diff
@@ -0,0 +1,555 @@
+        # Parse the XML file
+        try:
+            xmltree = ET.parse(resultsfile)
+        except FileNotFoundError as err:
```
Use `tidy3d.exceptions.FileError` instead.

yaugenst-flex commented on tidy3d/plugins/klayout/drc/drc.py:
```diff
@@ -0,0 +1,555 @@
+                "-rd",
+                f"resultsfile={self.resultsfile}",
+            ]
+        )
```
There needs to be some error handling here as subprocess can be brittle. At the very least we should check for status codes.

yaugenst-flex commented on tidy3d/plugins/klayout/drc/drc.py:
```diff
@@ -0,0 +1,555 @@
+        title="DRC Results File",
+        description="Path to the KLayout DRC results file.",
+    )
+
```
Should add path validators to check for, e.g., valid extensions for all input files (`gdsfile`, `designrulefile`, `resultsfile`). Validators can also check for existence of files (if required), which would obviate the need for the below helper functions.

yaugenst-flex commented on tidy3d/plugins/klayout/drc/drc.py:
```diff
@@ -0,0 +1,555 @@
+        """
+        # Parse the XML file
+        try:
+            xmltree = ET.parse(resultsfile)
```
This is more of a "maybe consider in the future": I don't know how large these files can get but I assume for large layouts quite big. Maybe these should be optimized to not load the entire file into memory at some point.

yaugenst-flex reviewed (changes requested):
Thanks @bzhangflex, lots of changes! Some nice to haves to consider:
1. I still think default paths would improve first-time usability, especially for the higher-level object-oriented interface.
2. Might be nice to add some logging (and maybe progress reporting?).

groberts-flex commented on tidy3d/plugins/klayout/drc/drc.py:
```diff
@@ -0,0 +1,555 @@
+        title="DRC Results File",
+        description="Path to the KLayout DRC results file.",
+    )
+    violations: dict[str, DRCViolation] = pd.Field(
```
what are the possible violation categories? was thinking this might be a dict that maps str -> tuple of DRCViolation's all for that category

bzhangflex replied:
Infinitely many haha, category is the same as "rule name" so anything the user decides to name their rules. The way it is currently DRCViolation.violations is a tuple of all the violation shapes, so if a user wants all the shapes for a particular category they would call DRCResults[category].violations.

groberts-flex replied:
gotcha, thanks for explanation! would it actually `DRCResults.violations[category]`? in which case wouldn't it be a type of `dict[str, tuple[DRCViolation, ...]]`?

bzhangflex replied:
It would be DRCResults.violations[category].violations but I overrode the __getitem__() for DRCResults so you can index it like a dict. Maybe it's confusing to have two levels of `violations` though, I could rename the inner one to `markers` or something like that

groberts-flex replied:
oh I see - sorry taking me a little to get there on this one! so there are categories and inside each category there are different types of violations possible? I originally was thinking of violation and category as the same thing. Maybe it would make sense to rename the outer one actually to something other than violations since it's a rule category? Although I can see your reasoning for renaming the inner one too. Overall, good with either, but I think it could be easier to follow if they had different names.

bzhangflex replied:
Thanks @groberts-flex for the suggestion! I've renamed the outer one to `DRCResults.violations_by_category` and the inner to `DRCViolation.markers`, let me know what you think.

groberts-flex commented on tidy3d/plugins/klayout/drc/drc.py:
```diff
@@ -0,0 +1,555 @@
+        return PolygonMarker(polygon=tuple(coords))
+
+    @staticmethod
+    def parse_violation_value(value: str) -> Union[EdgeMarker, EdgePairMarker, PolygonMarker]:
```
is this parsing the string output from a `DRCResults` object? is there a case where this is needed instead of just using the DRCResults object directly?

bzhangflex replied:
It's parsing the XML file that KLayout generates after running DRC. It's a good point though, I think these parsing functions can be moved out of the DRCResults object entirely

yaugenst-flex replied:
yes that sounds great, and they can also in principle be used  (and tested!) independently

tylerflex commented on tidy3d/plugins/klayout/drc/README.md:
```diff
@@ -0,0 +1,106 @@
+
+1. Have KLayout installed and added to your system PATH
+
+To check if KLayout has been added to the system PATH, you can use the `shutil.which` utility in Python:
```
very minor suggestion but, I wonder if we could wrap this into a quick utility function like `tidy3d.plugins.klayout.check_installation()`

tylerflex commented on tidy3d/plugins/klayout/drc/README.md:
```diff
@@ -0,0 +1,106 @@
+
+## Prerequisites
+
+1. Have KLayout installed and added to your system PATH
```
it would be good to also link to instructions on how to do this (or even a quick code snippet)

tylerflex commented on tidy3d/plugins/klayout/drc/README.md:
```diff
@@ -0,0 +1,106 @@
+from tidy3d.plugins.klayout.drc import run_drc
+
+# Run DRC on a Tidy3D object
+results = run_drc(
```
would be nice to include keywords here too I think.

bzhangflex replied:
Thanks @tylerflex! What do you mean by keywords?

tylerflex replied:
just the argument names like `run_drc(arg1=blah, ...)`

bzhangflex replied:
Ah got it

tylerflex reviewed (commented):
This is great. simple, useful, and decoupled enough from klayout implementation that it seems easily extendable. As a klayout newbie, I just had a few minor comments about the README to help improve context a bit. I think also a complete, simple, Quickstart example might be useful if that's possible. Either a separate directory `example/` in the plugin and/or a section in the README with code snippets for each of the files and a simple instruction.

Thanks @bzhangflex !

bzhangflex commented:
Thanks @yaugenst-flex, @groberts-flex and @tylerflex! I've pushed a latest version with the following changes, please let me know what you think:

- Refactored the codebase so the abstraction hierarchy has the higher level classes wrapping pure functions.
- Separated out DRC running and result into separate modules for better organization
- Added fixes and improvements suggested by all
- Updated documentation and tests
- Added a quickstart example under plugins/klayout/drc/example

tylerflex commented on tidy3d/plugins/klayout/drc/README.md:
```diff
@@ -0,0 +1,131 @@
+
+To install KLayout, please refer to https://www.klayout.de/build.html. 
+
+After installation, the application may need to be manually added to the system PATH. The method to add KLayout to your system PATH depends on your operating system. For example, on MacOS, you can add the following line to your `~.zshrc` file. This will permanently add KLayout to your PATH:
```
`~/.zshrc` I think?

yaugenst-flex replied:
TIL that `zsh` is default in macos...

tylerflex commented on tidy3d/plugins/klayout/drc/README.md:
````diff
@@ -0,0 +1,131 @@
+```python
+from tidy3d.plugins.klayout import check_installation
+# Prints the full path to the executable if found, otherwise returns None
+print(check_installation())
````
this works, I guess another way would be for the function to just return if it's working or otherwise raise an error

bzhangflex replied:
There is [this function](https://github.com/flexcompute/tidy3d/blob/f8d577d3fa6e6d520621dd00549e53660dde818c/tidy3d/plugins/klayout/util.py#L21) that raises an error if klayout isn't found. It gets called automatically when the user imports the drc module, does that seem like a good solution?

tylerflex replied:
I think that should be fine

tylerflex commented on tidy3d/plugins/klayout/drc/drc.py:
```diff
@@ -0,0 +1,310 @@
+    def check_gdsfile_filetype(cls, v: Path) -> Path:
+        """Check GDS filetype is .gds"""
+        if v.suffix != ".gds":
+            raise ValidationError(f"GDS file '{v}' must end with .gds")
```
let's put `.gds` int he single quotes

tylerflex commented on tidy3d/plugins/klayout/drc/drc.py:
```diff
@@ -0,0 +1,310 @@
+    def check_designrulefile_filetype(cls, v: Path) -> Path:
+        """Check DRC filetype is .drc"""
+        if v.suffix != ".drc":
+            raise ValidationError(f"DRC file '{v}' must end with .drc")
```
let's also put `.drc` in the singe quotes.

tylerflex commented on tidy3d/plugins/klayout/drc/drc.py:
```diff
@@ -0,0 +1,310 @@
+    def check_designrulefile_exists(cls, v: Path) -> Path:
+        """Check DRC file exists"""
+        if not _check_file_exists(v):
+            raise ValidationError(f"DRC file '{v}' does not exist")
```
make sure you add periods to the end of the validation errors

tylerflex commented on tidy3d/plugins/klayout/drc/drc.py:
```diff
@@ -0,0 +1,310 @@
+        >>> print(results)
+        """
+        return run_drc(
+            td_object,
```
as a general rule, let's use keywords to pass these things around. eg `a=b` in the function. it's easier to see what these are passing rather than relying on order

tylerflex replied:
this pops up in a few other places.

tylerflex commented on tidy3d/plugins/klayout/drc/drc.py:
```diff
@@ -0,0 +1,310 @@
+from tidy3d.plugins.klayout.util import _check_file_exists, _check_klayout_on_path
+
+
+class DRCRunner(Tidy3dBaseModel):
```
the DRCRunner seems like a convenient way to call these other functions. I wonder if it makes sense to just move eg. `def run_drc_on_gds` `def run_drc` and the validator functions into this class as methods, rather than have them floating. Maybe @yaugenst-flex has some opinions about this.

Could also stick the other options such as `verbose` into fields of `DRCRunner`. I think this might be a preferable design overall. would be curious to hear what others think.

bzhangflex replied:
@yaugenst-flex had suggested to have `run_drc_on_gds` and `run_drc` as standalone functions and to wrap them with a `DRCRunner` class for higher level abstraction. I think it makes sense because `run_drc_on_gds` and `run_drc` don't need to modify state, and decoupling them from a class makes it easier to reuse them elsewhere. 

Perhaps the validator functions could be moved into `DRCRunner` but then the standalone functions wouldn't be able to make use of them, so I'm not sure what's best there. Would like to hear what you think.

`verbose` could be added as a `DRCRunner` field. Would we want to make `td_object` a `DRCRunner` field too? I'm not sure if `**to_gds_file_kwargs` could be added though since they are keyword args?

groberts-flex replied:
are there cases where `run_drc_on_gds` and `run_drc` would need to be completely standalone from the `DRCRunner` class (i.e. - they need to be called but it is inconvenient to first make the `DRCRunner` class)?

maybe this is how you had it before, but the DRCRunner could also just be the argument to run_drc or run_drc_on_gds (maybe then it would have a different name)? it would essentially be a way to define all the arguments to the function such that you were confident they had been validated upon calling the function. and then the functions could return a DRCResults?

bzhangflex replied:
> are there cases where run_drc_on_gds and run_drc would need to be completely standalone from the DRCRunner class (i.e. - they need to be called but it is inconvenient to first make the DRCRunner class)?

perhaps if we wanted to make a more general DRCRunner class that interfaces with other tools in the future? Not sure how strong of an argument that is for standalone `run_drc` functions though.

> maybe this is how you had it before, but the DRCRunner could also just be the argument to run_drc or run_drc_on_gds (maybe then it would have a different name)? 

I think the idea is that `DRCRunner` could be extended in the future with additional functionality so it should live higher in the abstraction hierarchy and call the lower level functions. Maybe validation could be moved out of the lower level functions and left only to `DRCRunner`, but I like the idea of having some validation within the lower level functions. Curious what others think.

yaugenst-flex replied:
Yes I am against adding all the logic into a monolithic `DRCRunner` class. I think, whenever possible, these functions should be stateless and side-effect free (well to be fair they can't be side-effect free because they call KLayout but hey..). That makes testing easier and more robust (don't have to create or mock the whole class), and you can expose these functions to users too. The class can then add type checks, defaults, validation, and other convenience methods that add "sugar" on top of those engines. We could consider adding a `DRCConfig` or `DRCSpec` to those functions (and the class) though, which would make the signature more stable, i.e., easier to extend over time.

yaugenst-flex replied:
That being said though, I'm not sure about the ergonomics of `DRCRunner.run_drc(args)` and `DRCRuner.run_drc_on_gds()` - it seems strange to have two "run DRC" methods, with one not taking arguments (and hence being a sort of "first class citizen") and the other one where you have to explicitly pass a tidy3d object. I think:
1. The runner should not have a `gdsfile` field - I don't think any of its functionality is tied to running DRC on a _specific_ file.
2. There should only be one `DRCRunner.run_drc()` (or even just `.run()`) function that takes as an argument either a tidy3d object or a gds file and then delegates to the right function to call.

tylerflex replied:
yea I agree with what yannick is saying 

But just so I'm clear, the fundamental function doing all the work is just this right?

```py
   output = subprocess.run(
        [
            "klayout",
            "-b",
            "-r",
            designrulefile,
            "-rd",
            f"gdsfile={gdsfile}",
            "-rd",
            f"resultsfile={resultsfile}",
        ],
        capture_output=True,
    )
```
It's essentially converting a GDS file + a design rule file to a results file?

And then everything else is essentially either 
b) converting user input to and from this? 
a) validating the inputs 

I think maybe one strategy is to simply capture all of the "static" parameters in a config, such as the results filename, design rule, verbosity or other options.

and then expose some methods that run this config on either tidy3d objects or GDS file (like yannick said I think)

Does that make sense or is my brain just totally fried at the end of the day? :D 

BTW I think we could then expose both a `DRCConfig.run("file.gds")` or a some standalone functions like

```py
def run_drc_on_gds(fname, **kwargs):
     config = DRCConfig(**kwargs) # validates everything
     return config.run(fname)
```

tylerflex replied:
so in summary, this could be one approach (pseudocode)

```py
def _run_drc(designrulefile, rdsfile, results_file):
   return subprocess.run(
        [
            "klayout",
            "-b",
            "-r",
            designrulefile,
            "-rd",
            f"gdsfile={gdsfile}",
            "-rd",
            f"resultsfile={resultsfile}",
        ],
        capture_output=True,
    )
    
class DRCConfig:
    designrule_fname : str
    results_fname : str
    verbose : bool = True
    capture_output: bool = True # for example

    def run(self, input: str | td.Simulation) -> None:
         """ wraps the pure function """
         if isinstance(input, td.Simulation):
              # convert to gds first
         return _run_drc(self.designrule_fname, gds_file, self.results_fname)

def run_drc(fname, **kwargs):
    """ user facing"""
     config = DRCConfig(**kwargs) # validates everything
     return config.run(fname)


```

yaugenst-flex replied:
Not quite, I was thinking the `DRCConfig` would be much more similar to `DerivativeInfo`. Essentially a pure dataclass that holds (and possibly validates) the parameters, with `run_drc_on_gds(drc_config: DRCConfig)`.

The `DRCRunner` would be a higher-level class that holds persistent fields (such as verbosity, design rules..) while the `DRCConfig` would be the config for each particular run, e.g. in this case it would also have a `gds_file` field. `DRCConfig` takes care of validation so that `run_drc_on_gds` can then assume that the inputs are valid.

So:
```py
runner = DRCRunner(rules="rules.ext", verbose=True)
runner.run("layout.gds")

# inside runner:
def run(self, gds_filepath):
  config = DRCConfig(rules=self.rules, verbose=self.verbose, gds_file=gds_filepath)
  return run_drc_on_gds(config)
```

tylerflex replied:
ok yea, I was also thinking about this strategy. I feel like that could work too.

bzhangflex replied:
Thanks @tylerflex and @yaugenst-flex, I've refactored the codebase with this structure. Now `run_drc_on_gds` is the base pure function that calls the DRC subprocess, a `DRCConfig` data class is used to hold the options and validate them, and `DRCRunner.run_drc` can run DRC on either a gds file or tidy3d object (calling `run_drc_on_gds` under the hood).

tylerflex commented on tidy3d/plugins/klayout/drc/README.md:
find myself wishing there was just a single list of functions that I could browse. But then again, if we stick them all as methods of the DRCRunner class, these methods would just come up in the API reference automatically, so that's another benefit of that approach.

bzhangflex replied:
I see, good point. I think there's only three user-facing functions intended to be used, which are `DRCRunner.run_drc` (or the lower level `run_drc`), `DRCRunner.run_drc_on_gds` (or the lower level `run_drc_on_gds`), and `load_drc_results`. Other functions are validators or helper functions not really meant for end users. I was hoping the Features section of the README would clarify that (realized I need to update it to use the class functions though). Let me know what you think.

tylerflex replied:
I think adding class methods to the features section of the README is probably good enough. thanks

groberts-flex commented on tidy3d/plugins/klayout/drc/drc.py:
```diff
@@ -0,0 +1,310 @@
+def _check_gdsfile_exists(gdsfile: Union[str, Path]) -> None:
+    """Check if the GDS file exists.
+
+    Parameters
```
missing gdsfile here in parameters description

groberts-flex commented on tidy3d/plugins/klayout/drc/drc.py:
```diff
@@ -0,0 +1,310 @@
+        raise FileNotFoundError(f"GDS file not found: '{gdsfile}'")
+
+
+def _check_designrulefile_exists(designrulefile: Union[str, Path]) -> None:
```
should there just be one _check_file_exists function and maybe pass a custom string for error message?

bzhangflex replied:
good idea, thanks. That will simplify the code

yaugenst-flex commented on tidy3d/plugins/klayout/drc/drc.py:
```diff
@@ -0,0 +1,310 @@
+
+    # run drc
+    if verbose:
+        print(
+            f"Running KLayout DRC on GDS file '{gdsfile}' with runset '{designrulefile}' and saving results to '{resultsfile}'..."
+        )
```
Replace all these `print`s with `td.log`

bzhangflex replied:
I updated this, but it no longer prints to the console. Is that intended or is something going wrong?

bzhangflex replied:
Nevermind now it's working

yaugenst-flex commented on tidy3d/components/types.py:
```diff
@@ -261,3 +261,9 @@ def __modify_schema__(cls, field_schema):
+
+DRCEdge = tuple[Coordinate2D, Coordinate2D]
+DRCEdgePair = tuple[DRCEdge, DRCEdge]
+DRCPolygon = tuple[Coordinate2D, ...]
```
not bad advice but seems like overdoing it

yaugenst-flex commented on tidy3d/plugins/klayout/drc/__init__.py:
```diff
@@ -0,0 +1,21 @@
+from .results import DRCResults, load_drc_results
+
+# Check if klayout exists in system path
+_check_klayout_on_path()
```
this is a plugin and the plugin is not imported by default (import tidy3d), and is unusable if klayout is not present. hence, a check at import time makes sense

yaugenst-flex commented on CHANGELOG.md:
```diff
@@ -40,6 +40,7 @@ with fewer layers than recommended.
 - Added heat sources with custom spatial dependence. It is now possible to add a `SpatialDataArray` as the `rate` in a `HeatSource`.
 - Added Transient Heat simulations. It is now possible to run transient Heat simulations. This can be done by specifying `analysis_spec` of `HeatChargeSimulation` object as `UnsteadyHeatAnalysis`. 
 - A `num_grid_cells` field to `WavePort`, which ensures that there are 5 grid cells across the port. Grid refinement can be disabled by passing `None` to `num_grid_cells`.
+- Added KLayout plugin, with DRC functionality for running design rule checks in `plugins.klayout.drc`. Supports running DRC on `Geometry`, `Structure`, and `Simulation` objects.
```
that's literally the case

bzhangflex commented on tidy3d/plugins/klayout/drc/results.py:
```diff
@@ -0,0 +1,337 @@
+    coords_pattern = r"([\d.-]+),([\d.-]+)"  # match the coordinates
+
+    # Extract the outer polygon coordinates
+    outer_match = re.match(outer_pattern, value.split("/")[0])
```
Good catch

bzhangflex commented on tests/test_plugins/klayout/drc/drc_results.lyrdb:
```diff
@@ -0,0 +1,82 @@
+   <comment/>
+   <image/>
+   <values>
+    <value>edge-pair: (-1.717,-0.141;-2,0)|(-2,0;-1.776,0.224)</value>
```
this is fine

bzhangflex commented on tidy3d/plugins/klayout/drc/results.py:
```diff
@@ -0,0 +1,337 @@
+
+        # Initialize violations dict with all the categories
+        violations = {}
+        for category in xmltree.getroot().findall(".//categories/category/name"):
+            violations[category.text] = DRCViolation(category=category.text)
```
I think this is fine, the tree isn't searched multiple times

bzhangflex commented on tidy3d/plugins/klayout/drc/drc.py:
```diff
@@ -0,0 +1,316 @@
+        capture_output=True,
+    )
+
+    if output.returncode != 0:
+        raise RuntimeError(f"KLayout DRC failed with error message: '{output.stderr}'.")
```
probably fine

bzhangflex commented:
Thanks all! Here are the latest updates:

- Refactored `drc.py` with better code organization. `run_gds_on_drc` is the base pure function that calls the DRC subprocess, a `DRCConfig` class was made for passing and validating parameters to `run_gds_on_drc`, and `DRCRunner`'s fields and `run_drc` method have been updated for better ergonomics.
- For result parsing, there was an uncaught case (that greptile caught!) where a polygon result could actually contain multiple polygons (eg if the result is from a hole in the layout). I've fixed this and replaced `PolygonMarker` with `MultiPolygonMarker` that contains a tuple of polygons.
- Many smaller code improvements
- Updated README, examples, tests, and docstrings to reflect changes
- Minor but I've renamed `designrulefile` to `drc_runset` since that is more standard terminology in EDA

yaugenst-flex commented on tests/test_plugins/klayout/drc/test_drc.py:
```diff
@@ -0,0 +1,328 @@
+filepath = Path(os.path.dirname(os.path.abspath(__file__)))
+
+
+def test_check_klayout_not_installed():
+    """Test that _check_klayout_on_path raises RuntimeError if klayout is not in the system path.
+    Assumes this is the case on the CI machine.
+    """
+    with pytest.raises(RuntimeError):
+        _check_klayout_on_path()
```
So this is true, but this is actually the first plugin I think where haven't added the dependency in `pyproject.toml`. Should we? _Can_ we?

bzhangflex replied:
It would have to install the KLayout binary, I'm guessing that's not really possible or recommended?

yaugenst-flex replied:
Yeah I saw [this](https://pypi.org/project/klayout/) but that doesn't do what we need. So yeah it's fine as is

yaugenst-flex commented on tests/test_plugins/klayout/drc/test_drc.py:
```diff
@@ -0,0 +1,328 @@
+
+    def test_drc_result_markers(self, drc_results):
+        """Test that the DRC result markers are correct"""
+        assert drc_results["min_width"].markers[0].edge_pair[0] == ((-0.6, 0.163), (-0.6, 0.419))
```
How were these values obtained? By running KLayout on them? Would these tests be invalid if e.g. KLayout changed some part of their DRC implementation? If so, this seems like "overtesting".

bzhangflex replied:
Yes `drc_results` are the output of a DRC run that I ran manually and uploaded the results. Ideally we would run KLayout DRC on the CI machines but I'm not sure if we can or want to do that?

yaugenst-flex commented on tidy3d/components/types.py:
```diff
@@ -261,3 +261,10 @@ def __modify_schema__(cls, field_schema):
+DRCEdge = tuple[Coordinate2D, Coordinate2D]
+DRCEdgePair = tuple[DRCEdge, DRCEdge]
+DRCPolygon = tuple[Coordinate2D, ...]
+DRCMultiPolygon = tuple[DRCPolygon, ...]
```
These should be moved to the plugin

yaugenst-flex commented on tidy3d/plugins/klayout/__init__.py:
```diff
@@ -0,0 +1,14 @@
+    "check_installation",
+    "load_drc_results",
+    "run_drc_on_gds",
+]
```
I think "DRCConfig", "DRCResults", "DRCRunner" should also be exposed here?

bzhangflex replied:
Yes good catch, thanks!

yaugenst-flex commented on tidy3d/plugins/klayout/drc/__init__.py:
```diff
@@ -0,0 +1,28 @@
+from .results import DRCResults, load_drc_results
+
+# Check if klayout exists in system path
+_check_klayout_on_path()
```
This check should run at the KLayout plugin level, not DRC.

bzhangflex replied:
I placed this check here instead of at the KLayout plugin level because at the KLayout plugin level, we could integrate with their Python API which doesn't require the full app to be installed + added to path.

yaugenst-flex replied:
Ok I see. I guess it makes sense then if the plugin is usable without full klayout installed

yaugenst-flex commented on tidy3d/plugins/klayout/drc/drc.py:
```diff
@@ -0,0 +1,230 @@
+        title="GDS File",
+        description="The path to the GDS file to write the Tidy3D object to.",
+    )
+    drc_runset: Path = pd.Field(
```
You could use `pydantic.FilePath` here, which will automatically validate that the path exists and is a file.

yaugenst-flex commented on tidy3d/plugins/klayout/drc/drc.py:
```diff
@@ -0,0 +1,230 @@
+    )
+
+    @validator("gdsfile")
+    def check_gdsfile_exists(cls, v: Path) -> Path:
```
Prefer `_validate_XXX` naming scheme for validators

yaugenst-flex commented on tidy3d/plugins/klayout/drc/drc.py:
```diff
@@ -0,0 +1,230 @@
+        self,
+        gdsfile: Path,
+        resultsfile: Path,
+        td_object: Optional[Union[Geometry, Structure, Simulation]] = None,
```
I think this function should only have to arguments: the input on which to run DRC, and (optionally) the file to write results to, with a default. So ideally you would just call `.run_drc(sim)` or `.run_drc(path)`. The method then dispatches to the appropriate function.

yaugenst-flex replied:
Also, since the class is already named `DRCRunner`, maybe it's nicer to just call this `run`?

yaugenst-flex commented on tidy3d/plugins/klayout/drc/drc.py:
```diff
@@ -0,0 +1,230 @@
+    """Configuration for KLayout DRC."""
+
+    gdsfile: Path = pd.Field(
+        default="geom.gds",
```
This class should not have defaults, it's just a container.

yaugenst-flex commented on tidy3d/plugins/klayout/drc/README.md:
```diff
@@ -0,0 +1,138 @@
+
+This module provides integration between Tidy3D and KLayout's Design Rule Check (DRC) engine, allowing you to perform design rule checks on GDS files and Tidy3D objects.
+
+## Quickstart
+
+For a full quickstart example script, please see [tidy3d/plugins/klayout/drc/example/example.py](./example/example.py).
```
I think it's great in principle that we have an example but we have to think about packaging, because these examples will be buried somewhere in `site-packages/`. I think instead we should provide this in the form of a notebook in the notebooks repo and reference it from here? Otherwise these are not really discoverable.

bzhangflex replied:
Yes definitely. Would we need to merge this first though before adding a notebook?

yaugenst-flex replied:
Actually notebook first :smile: and then update the submodule in this repo to point to the correct notebook repo state

yaugenst-flex replied:
As long as it's done before release though it doesn't matter too much

yaugenst-flex commented on tidy3d/plugins/klayout/drc/results.py:
```diff
@@ -0,0 +1,364 @@
+    category: str = pd.Field(
+        title="DRC Violation Category", description="The category of the DRC violation."
+    )
+    markers: tuple[Union[EdgeMarker, EdgePairMarker, MultiPolygonMarker], ...] = pd.Field(
```
Could make sense to introduce a new type like `DRCMarker`?

yaugenst-flex commented on tidy3d/plugins/klayout/drc/results.py:
```diff
@@ -0,0 +1,364 @@
+        title="DRC Violation Category", description="The category of the DRC violation."
+    )
+    markers: tuple[Union[EdgeMarker, EdgePairMarker, MultiPolygonMarker], ...] = pd.Field(
+        default=(), title="DRC Markers", description="Tuple of DRC markers in this category."
```
Should this have a default?

bzhangflex replied:
I suppose it doesn't need it.

yaugenst-flex replied:
Basically I was wondering when it would ever make sense to have an empty `DRCViolation` instance? If that's not a thing, then `markers` should be a required argument.

bzhangflex replied:
Ah I see. I think it's useful to have empty `DRCViolation`s for rule bookkeeping, so the user can see all of the rules in `DRCResults`, not just ones that have been flagged

yaugenst-flex commented on tidy3d/plugins/klayout/drc/results.py:
```diff
@@ -0,0 +1,364 @@
+class DRCResults(Tidy3dBaseModel):
+    """A class for loading and storing KLayout DRC results."""
+
+    resultsfile: Path = pd.Field(
```
Probably also should be a `pydantic.FilePath`?

yaugenst-flex commented on tidy3d/plugins/klayout/drc/results.py:
```diff
@@ -0,0 +1,364 @@
+    """A class for loading and storing KLayout DRC results."""
+
+    resultsfile: Path = pd.Field(
+        default="drc.lyrdb",
```
I already forgot, but is this the default path that the results get written to? Need to make sure that that is consistent.

yaugenst-flex replied:
Actually, we should probably define these as module- or plugin-level constants.

yaugenst-flex commented on tidy3d/plugins/klayout/drc/results.py:
```diff
@@ -0,0 +1,364 @@
+    )
+
+    @cached_property
+    def clean(self) -> bool:
```
Maybe `is_clean`?

yaugenst-flex commented on tidy3d/plugins/klayout/drc/results.py:
```diff
@@ -0,0 +1,364 @@
+        return summary
+
+    @classmethod
+    def from_file(cls, resultsfile: Union[str, Path]) -> DRCResults:
```
Wait, upon re-reading this class, I think the fields seem redundant. Shouldn't you initialize _either_ with a results file _or_ with the already-parsed violations? It doesn't really make sense to have both I think? I think the `DRCResults` should only have one field (the violations), and a constructor `.from_file(path)` that parses the files and constructs the `DRCResults` from that.

yaugenst-flex replied:
Oh but be careful! Actually we can't name it `.from_file` because that's already the `Tidy3dBaseModel` method to load a serialized model. So should be `.from_results_file()` or similar.

yaugenst-flex commented on tidy3d/plugins/klayout/drc/results.py:
```diff
@@ -0,0 +1,364 @@
+        return cls(resultsfile=resultsfile, violations_by_category=violations)
+
+
+def load_drc_results(resultsfile: Union[str, Path]) -> DRCResults:
```
Also a case of dependency inversion here. I think the logic to parse a file into DRC violations should move here, with a signature like `parse_drc_file(path) -> dict[str, DRCViolation]`

bzhangflex replied:
Should we keep the `load_drc_results()` alias or only use `DRCResults.from_results_file()`?

yaugenst-flex replied:
Up to you. Personally I think having a function with the signature I mentioned above makes sense, but on the other hand this likely won't ever be used outside of the class context so it might make more sense to have it in the class.

bzhangflex replied:
Ah I meant having a function `parse_drc_file(path) -> dict[str, DRCViolation]` that will be used by `DRCResults.from_results_file()`, along with a `load_drc_results() -> DRCResults` that just returns `DRCResults.from_results_file()`

yaugenst-flex replied:
Ah no both of those do the same thing then no? Just one `DRCResults.from_results_file()` seems correct? Although, is `DRCResults.load()` maybe nicer? Just a naming thing.

yaugenst-flex commented on tidy3d/plugins/klayout/util.py:
```diff
@@ -0,0 +1,52 @@
+from typing import Union
+
+
+def check_installation() -> Union[str, None]:
+    """Checks if KLayout is installed and added to the system PATH.
+    If it is, this returns the path to the executable. Otherwise, returns ``None``.
+    Equivalent to $which("klayout") in the terminal.
+
+    Returns
+    -------
+    str
+        The path to the KLayout executable. If KLayout is not found, returns ``None``.
+    """
+    return which("klayout")
+
+
+def _check_klayout_on_path() -> None:
+    """Raises an error if KLayout is not found in the system path.
+
+    Raises
+    ------
+    RuntimeError
+        If KLayout is not found in system PATH.
+    """
+    # Check if klayout exists in system path
+    if check_installation() is None:
+        raise RuntimeError(
+            "KLayout was not found in the system PATH. Please ensure KLayout is installed and added to your system PATH before using the KLayout DRC plugin."
+        )
```
Is there a reason to split this up into two functions? It seems like `check_installation()` should raise, according to its name? Why is `_check_klayout_on_path` a "private" helper that calls a public (`check_installation()`) function?

bzhangflex replied:
@tylerflex suggested having a `check_install` function for users. I thought it would be better for the user facing function to return the KLayout install path rather than error. `_check_klayout_on_path` is called on module import to enforce KLayout installation

yaugenst-flex replied:
I see. Maybe we could do `check_installation(raise: bool = False)` and have our check simply run with `raise=True`? And otherwise maybe we print a warning if KLayout isn't found?

bzhangflex replied:
That's a good idea, thanks!

bzhangflex replied:
It seems like `_check_klayout_on_path` breaks the CI tests because it raises when the drc functions are imported 😅. Wonder what we should do here

yaugenst-flex replied:
Oooh right this is a problem. Not only for CI. Tests will fail if you don't have KLayout installed 🤔

So actually I think we should reconsider the failure on `__init__`. The _only_ function that actually requires KLayout to be installed is `run_drc_on_gds`, so let's just move the check there? Then we can monkeypatch this function in the tests (replace with a dummy that just loads the `DRCResults`) and everything should work?

bzhangflex commented:
Thanks @yaugenst-flex! Just pushed an update and will work on a notebook example

yaugenst-flex commented on tidy3d/plugins/klayout/drc/example/example.py:
This needs to be removed (the whole folder actually)

yaugenst-flex commented on tidy3d/plugins/klayout/drc/drc.py:
```diff
@@ -0,0 +1,225 @@
+class DRCConfig(Tidy3dBaseModel):
+    """Configuration for KLayout DRC."""
+
+    gdsfile: pd.FilePath = pd.Field(
```
Have you tested whether this works? I thought `FilePath` validates that the path _must_ exist? This is the difference between `Path` and `FilePath`. See https://docs.pydantic.dev/1.10/usage/types/#pydantic-typeshttps://docs.pydantic.dev/1.10/usage/types/#pydantic-types

bzhangflex replied:
The config is only used for `run_drc_on_gds()` so it should validate the gds file exists.

yaugenst-flex replied:
ooh right

yaugenst-flex commented on tidy3d/plugins/klayout/drc/drc.py:
```diff
@@ -0,0 +1,225 @@
+        title="DRC Runset file",
+        description="Path to the KLayout DRC runset file.",
+    )
+    resultsfile: Path = pd.Field(
```
Should be pydantic `Path`

bzhangflex replied:
It doesn't seem like pydantic has a `Path` type from what I can tell

yaugenst-flex replied:
huh indeed, seems like it's a pydantic v2 thing

yaugenst-flex reviewed (commented):
Thanks @bzhangflex code looks great, no further comments. We can merge once the documentation/examples are figured out!

yaugenst-flex commented on tidy3d/plugins/klayout/drc/drc.py:
```diff
@@ -0,0 +1,225 @@
+
+    def run(
+        self,
+        td_object_or_gdsfile: Union[Geometry, Structure, Simulation, Path],
```
Oh actually one comment: While this is descriptive, it feels a bit hard to use. Maybe something simpler like `source`?

yaugenst-flex reviewed (commented):
Eventually we should expand `DRCResults` to also be able to export the violations into some nice format for visualization. Like putting the violations into some gds layers.

yaugenst-flex commented on CHANGELOG.md:
```diff
@@ -76,6 +76,7 @@ with fewer layers than recommended.
 - Added heat sources with custom spatial dependence. It is now possible to add a `SpatialDataArray` as the `rate` in a `HeatSource`.
 - Added Transient Heat simulations. It is now possible to run transient Heat simulations. This can be done by specifying `analysis_spec` of `HeatChargeSimulation` object as `UnsteadyHeatAnalysis`. 
 - A `num_grid_cells` field to `WavePort`, which ensures that there are 5 grid cells across the port. Grid refinement can be disabled by passing `None` to `num_grid_cells`.
+- Added KLayout plugin, with DRC functionality for running design rule checks in `plugins.klayout.drc`. Supports running DRC on `Geometry`, `Structure`, and `Simulation` objects.
```
This needs to move up into the "Unreleased" section

yaugenst-flex reviewed (approved):
LGTM! Changelog needs an edit but otherwise good to go.

tylerflex commented on tidy3d/plugins/klayout/drc/README.md:
either this could be the main readme under `tidy3d/plugins/klayout/README.md`, or that file should probably include a link to this readme so users know that it exists.

tylerflex commented on tidy3d/plugins/klayout/drc/README.md:
```diff
@@ -0,0 +1,134 @@
+
+## Quickstart
+
+For a full quickstart example, please see [this quickstart notebook](link/to/notebook).
```
note for now this link is broken

tylerflex commented on tidy3d/plugins/klayout/drc/drc.py:
```diff
@@ -0,0 +1,225 @@
+
+    Example
+    -------
+    Running DRC on a GDS file:
```
not sure how this will render in docs, might want to just comment out the titles like
```
>>> # Running DRC on a GDS File
>>> from tidy3d...

tylerflex commented on tidy3d/plugins/klayout/drc/results.py:
```diff
@@ -0,0 +1,364 @@
+    Parameters
+    ----------
+    value : str
+        The edge-pair value string from DRC result database, with format (x1,y1;x2,y2)|(x3,y3;x4,y4).
```
this (x1, ...) stuff might be better within double brackets ``.

tylerflex replied:
this pops up in various places (eg below). I'm just concerned it might be unreadable in the docs. (assuming this will go in the docs? If not maybe we can ignore.

tylerflex commented on tidy3d/plugins/klayout/drc/results.py:
```diff
@@ -0,0 +1,364 @@
+    Parameters
+    ----------
+    coords_str : str
+        The string of coordinates for a single polygon, with format 'x1,y1;x2,y2;...'.
```
double backslash brackets `` for docstrings

tylerflex reviewed (approved):
approving, a few really minor comments. Thanks @bzhangflex ! this will be great to have!

bzhangflex commented on tidy3d/plugins/klayout/drc/drc.py:
```diff
@@ -0,0 +1,224 @@
+        if isinstance(source, (Geometry, Structure, Simulation)):
+            gdsfile = td_object_gds_savefile
+            if self.verbose:
+                td.log.info(f"Writing Tidy3D object to GDS file '{gdsfile}'.")
```
@yaugenst-flex Should this (and others) just be `td.log(f...)` not `td.log.info(f..)`? I noticed this doesn't seem to print to console unless `td.logging_level` is set to `info`

yaugenst replied:
you always have to `td.log.some_level()`. `info` is the appropriate level here i would say.

bzhangflex replied:
ah sorry I got confused because I saw `console.log` used elsewhere and got that mixed up with `td.log`. Is `td.log` more appropriate here?

yaugenst replied:
where? that shouldn't happen 😄 `td.log`, always

bzhangflex replied:
like in the [web api](https://github.com/flexcompute/tidy3d/blob/eaf1f0a0912d7c108c492244d4a6d0c9a7e756ea/tidy3d/web/api/webapi.py#L270)

yaugenst-flex replied:
that's the `td.log` logging console 😉

yaugenst-flex replied:
but yeah you can probably also use that? i think that would always print, independent of the logging level?

yaugenst-flex commented:
@bzhangflex doctests are failing. once fixed we can merge: https://github.com/flexcompute/tidy3d/actions/runs/16772611050/job/47491169804#step:6:479

bzhangflex commented:
> @bzhangflex doctests are failing. once fixed we can merge: https://github.com/flexcompute/tidy3d/actions/runs/16772611050/job/47491169804#step:6:479

@yaugenst-flex It looks like its failing because of the DRC run examples, which can't be ran on the test machine without installing KLayout and providing additional files like a drc runset. What would you recommend here? I think it's good to have examples in the docstrings but maybe we can remove them and point to the quickstart notebook instead

tylerflex commented:
@bzhangflex just fyi you can skip the doctests using this special syntax
https://github.com/flexcompute/tidy3d/blob/f3d0f8e681bb5ba3eb75709c26ed2784df4671b6/tidy3d/web/api/tidy3d_stub.py#L120

```
>>> my_code.run() # doctest: +SKIP
```

bzhangflex commented:
> @bzhangflex just fyi you can skip the doctests using this special syntax
> 
> https://github.com/flexcompute/tidy3d/blob/f3d0f8e681bb5ba3eb75709c26ed2784df4671b6/tidy3d/web/api/tidy3d_stub.py#L120
> 
> ```
> >>> my_code.run() # doctest: +SKIP
> ```

Ah perfect thanks, didn't know about that. I'll add those in then.
